src/chainstate/stacks/index/trie_sql.rs

```diff
@@ -168,6 +235,64 @@ pub fn write_trie_blob<T: MarfTrieId>(
     Ok(block_id)
 }
 
+/// Write the offset/length of a trie blob that was stored to an external file.
+/// Do this only once the trie is actually stored, since only the presence of this information is
+/// what guarantees that the blob is persisted.
+pub fn write_external_trie_blob<T: MarfTrieId>(
+    conn: &Connection,
+    block_hash: &T,
+    offset: u64,
+    length: u64,
+    block_id: Option<u32>,
+) -> Result<u32, Error> {
+    let block_id = if let Some(block_id) = block_id {
+        // existing entry (i.e. a migration)
+        let empty_blob: &[u8] = &[];
+        let args: &[&dyn ToSql] = &[
+            block_hash,
+            &empty_blob,
+            &0,
+            &u64_to_sql(offset)?,
+            &u64_to_sql(length)?,
+            &block_id,
+        ];
+        let mut s =
+            conn.prepare("UPDATE marf_data SET block_hash = ?1, data = ?2, unconfirmed = ?3, external_offset = ?4, external_length = ?5 WHERE block_id = ?6")?;
+        s.execute(args)?;
+
+        debug!(
+            "Replaced block trie {} at rowid {} offset {}",
+            block_hash, block_id, offset
+        );
+        block_id
+    } else {
+        // new entry
+        let empty_blob: &[u8] = &[];
+        let args: &[&dyn ToSql] = &[
+            block_hash,
+            &empty_blob,
+            &0,
+            &u64_to_sql(offset)?,
+            &u64_to_sql(length)?,
+        ];
+        let mut s =
+            conn.prepare("INSERT INTO marf_data (block_hash, data, unconfirmed, external_offset, external_length) VALUES (?, ?, ?, ?, ?)")?;
+        let block_id = s
+            .insert(args)?
+            .try_into()
+            .expect("EXHAUSTION: MARF cannot track more than 2**31 - 1 blocks");
+
+        debug!(
+            "Wrote block trie {} to rowid {} offset {}",
+            block_hash, block_id, offset
+        );
+        block_id
+    };
```

Comment:
Is it possible for this method to be called _without_ a `block_id` if the `block_id` already exists in the database? Should this be part of the calling contract for this method (if so, this assumption should be made explicit in a rustdoc entry for the method)?

I guess this could also be made an `INSERT OR REPLACE`.

Reply:
I'll factor this into `write_external_trie_blob()` which stores a new trie blob, and `update_external_trie_blob()` which updates an existing trie blob.  They'll just call this method (which will be made into a private "inner" method) with the appropriate arguments.